### PR TITLE
feat(crit): add `crit x sk` for sockets

### PR DESCRIPTION
**Pull request by snprajwal**
This is useful for implementing socket views in checkpointctl, going forward.

### CODE REVIEW

File: crit/cli/cli.go [hunk 1]
```diff
@@ -242,14 +242,14 @@ var xCmd = &cobra.Command{
 		switch args[1] {
 		case "ps":
 			xData, err = c.ExplorePs()
-		case "fds":
+		case "fd":
 			xData, err = c.ExploreFds()
-		case "mems":
+		case "mem":
```
Comment (rst0git) [inline crit/cli/cli.go:247]: To preserve compatibility with the Python version of `crit`, should we accept both `mems` and `mem`?

File: crit/Makefile [hunk 2]
```diff
@@ -1,8 +1,10 @@
 GO ?= go
 CRIU ?= criu
+CRIT_SRC = $(shell find . -name \*.go)
```
Comment (rst0git) [inline crit/Makefile:3]:
```Suggestion
CRIT_SRC := $(shell find . -type f -name '*.go')
```

File: crit/explore.go [hunk 3]
```diff
@@ -145,6 +145,11 @@ func (c *crit) ExploreFds() ([]*Fd, error) {
 			Path: filePath,
 		})
 
+		// Omit if the process has no file descriptors
+		if len(fdEntry.Files) == 0 {
+			continue
+		}
```
Comment (rst0git) [inline crit/explore.go:151]: This change is not directly related to the renaming of `fds|mems` to `fd|mem`. It should be in a separate commit.

File: crit/Makefile [hunk 4]
```diff
@@ -1,8 +1,10 @@
...
+CRIT_CLI = cmd/main.go
```
[the remark this answers is not in the document]
Comment (snprajwal) [reply crit/Makefile:4]: I used `=` instead of `?=` since I felt it's reasonable to assume that we won't override this variable, unlike others. The CLI entrypoint will always be consistent, so I didn't see the need for `?=`.

[the remark this answers is not in the document]
Comment (rst0git) [reply crit/Makefile]: Sure, perhaps we should use [simple](https://www.gnu.org/software/make/manual/html_node/Simple-Assignment.html) (`:=`) instead of [recursive](https://www.gnu.org/software/make/manual/html_node/Recursive-Assignment.html) (`=`) assignment?

[the remark this answers is not in the document]
Comment (snprajwal) [reply crit/Makefile]: Changed and pushed in #141

File: crit/explore.go [hunk 5]
```diff
@@ -373,3 +373,122 @@ func (c *crit) ExploreRss() ([]*RssMap, error) {
...
+			socket := Socket{
+				Fd:   fdInfo.GetFd(),
+				Type: fdInfo.GetType().String(),
+			}
+			switch fdInfo.GetType() {
+			case fdinfo.FdTypes_INETSK:
+				file, err := getFile(c.inputDirPath, fdInfo.GetId())
+				if err != nil {
+					return nil, err
+				}
```
Comment (rst0git) [inline crit/explore.go:433]: It might be good to move this code just before the `switch` statement to avoid code duplication.
```Go
file, err := getFile(c.inputDirPath, fdInfo.GetId())
if err != nil {
	return nil, err
}
```

File: crit/explore.go [hunk 6]
```diff
@@ -373,3 +373,122 @@ func (c *crit) ExploreRss() ([]*RssMap, error) {
...
+// Sk represents the sockets associated with a single process
+type Sk struct {
+	PId     uint32    `json:"pId"`
+	Sockets []*Socket `json:"sockets"`
+}
+
+// Socket represents a single socket
+type Socket struct {
+	Fd       uint32 `json:"fd"`
+	Type     string `json:"type"`
```
Comment (rst0git) [inline crit/explore.go:386]: It might be better to use `FdType` here to avoid confusion between the fdInfo *type* and socket *type*.

File: crit/explore.go [hunk 7]
```diff
@@ -373,3 +373,122 @@ func (c *crit) ExploreRss() ([]*RssMap, error) {
...
+			switch fdInfo.GetType() {
+			case fdinfo.FdTypes_INETSK:
+				file, err := getFile(c.inputDirPath, fdInfo.GetId())
+				if err != nil {
+					return nil, err
+				}
+				if isk := file.GetIsk(); isk != nil {
+					socket.State = getSkState(isk.GetState())
+					socket.Family = getSkFamily(isk.GetFamily())
+					socket.Protocol = getSkProtocol(isk.GetType())
```
Comment (rst0git) [inline crit/explore.go:437]: It might be better to use `socket.Type` for the value of `GetType()` instead of `socket.Protocol` to avoid confusion.

For instance, socket types are `SOCK_STREAM = 1`,  `SOCK_DGRAM = 2`, `SOCK_RAW = 3`, etc. (see `man 2 socket`)

Comment (rst0git) [reply crit/explore.go:437]: For example, something like the following:

```Go
...
socket.Type = getSkType(isk.GetType())
...

func getSkType(skType uint32) string {
	switch skType {
	case syscall.SOCK_STREAM:
		return "STREAM"
	case syscall.SOCK_DGRAM:
		return "DGRAM"
	case syscall.SOCK_RAW:
		return "RAW"
	case syscall.SOCK_SEQPACKET:
		return "SEQPACKET"
	...
	default:
		return "unknown"
	}
}
```

File: crit/utils.go [hunk 8]
```diff
@@ -232,3 +244,97 @@ func getUnixSkFilePath(dir string, file *fdinfo.FileEntry, fID uint32) (string,
...
+		return "CLOSING"
+	case 12:
+		return "NEW_SYN_RECV"
+	default:
+		return ""
+	}
+}
+
+// Helper to identify socket family
+func getSkFamily(family uint32) string {
```
Comment (rst0git) [inline crit/utils.go:301]: `AF` is an abbreviation for "Address Family"
```Suggestion
func getAddressFamily(family uint32) string {
```

File: crit/explore.go [hunk 9]
```diff
@@ -373,3 +373,110 @@ func (c *crit) ExploreRss() ([]*RssMap, error) {
...
+type Sk struct {
+	PId     uint32    `json:"pId"`
+	Sockets []*Socket `json:"sockets"`
+}
+
+// Socket represents a single socket
+type Socket struct {
+	Fd       uint32 `json:"fd"`
+	FdType   string `json:"fdType"`
+	Family   string `json:"family,omitempty"`
```
Comment (rst0git) [inline crit/explore.go:387]: Would it make sense to also add `Protocol`? For example, something like the following: 
```Go
Protocol string `json:"proto,omitempty"`
```
```Go
socket.Protocol = getProto(isk.GetProto())
```
```Go
func getProto(proto uint32) string {
	switch proto {
	case syscall.IPPROTO_ICMP:
		return "ICMP"
	case syscall.IPPROTO_ICMPV6:
		return "ICMPV6"
	case syscall.IPPROTO_IGMP:
		return "IGMP"
	case syscall.IPPROTO_RAW:
		return "RAW"
	case syscall.IPPROTO_TCP:
		return "TCP"
	case syscall.IPPROTO_UDP:
		return "UDP"
	case syscall.IPPROTO_UDPLITE:
		return "UDPLITE"
	default:
		return ""
	}
}
```

Comment (snprajwal) [reply crit/explore.go:387]: Ah so protocol is shortened to `proto`. I mistook that to be the descriptor for the protobuf definition :sweat_smile:

File: crit/utils.go [hunk 10]
```diff
@@ -246,3 +258,129 @@ func (ps *PsTree) FindPs(pid uint32) *PsTree {
 	}
 	return nil
 }
+
+// Helper to convert slice of uint32 into IP address string
+func processIP(parts []uint32) string {
```
Comment (rst0git) [inline crit/utils.go:263]: Would it be possible to add unit tests for this function?

Comment (snprajwal) [reply crit/utils.go:263]: Yes that makes sense. I'll add it, and look into adding it for any other utility functions too if possible.

**Review by rst0git — APPROVED** [review]
LGTM
